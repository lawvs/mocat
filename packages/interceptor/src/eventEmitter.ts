// See https://github.com/EventEmitter2/EventEmitter2
import { EventEmitter2 } from 'eventemitter2'
import type { EventAndListener } from 'eventemitter2'
import type { MockEventMap, RegisterEvent, RunEvent } from './types'

export interface MockEventEmitter {
  emit<T extends keyof MockEventMap>(event: T, value: MockEventMap[T]): boolean

  on<T extends keyof MockEventMap>(
    event: T,
    listener: (value: MockEventMap[T]) => void
  ): void
  once<T extends keyof MockEventMap>(
    event: T,
    listener: (value: MockEventMap[T]) => void
  ): void
  onAny<T extends keyof MockEventMap>(
    listener: (event: T, value: MockEventMap[T]) => void
  ): void
  off<T extends keyof MockEventMap>(
    event: T,
    listener: (...values: any[]) => void
  ): this
  offAny(listener: (value: any[]) => void): this
}

// @ts-ignore
export const eventEmitter: MockEventEmitter = new EventEmitter2()

export const debug = () => {
  const log: EventAndListener = (event, ...payload) =>
    // eslint-disable-next-line no-console
    console.log(
      `%cReceive%c %c${event.toString()}`,
      'background: rgba(0, 255, 255, 0.6); color: black; padding: 0px 6px; border-radius: 4px;',
      '',
      'text-decoration: underline',
      payload
    )
  eventEmitter.onAny(log)
  return () => eventEmitter.offAny(log)
}

// Register event

// const mockMap = new Map<any, any>()

export const registerMock = (rule: RegisterEvent) => {
  // mockMap.set(target, { desc, scenes })
  eventEmitter.emit(rule.type, rule)
}

// Run event

export const onRun = (event: RunEvent) => {
  eventEmitter.emit(event.type, event)
}
