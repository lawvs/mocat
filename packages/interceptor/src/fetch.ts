// http://www.wheresrhys.co.uk/fetch-mock/
import FetchMock from 'fetch-mock'

import { onRun, matchNetworkRule } from './eventEmitter'
import { networkSceneToResponse, NOOP, passRequest } from './utils'
import type { NetworkBeforeEvent, NetworkScene } from './types'

let originalFetch: typeof fetch | null = null

export const realFetch = (...args: Parameters<typeof fetch>) =>
  (originalFetch || fetch)(...args)

const withResolveScene = (resolve: (response: Response) => void) => (
  scene: NetworkScene
) => resolve(networkSceneToResponse(scene))

export const setupFetch = () => {
  if (originalFetch) {
    console.warn('Already setup fetch！')
    return NOOP
  }
  originalFetch = globalThis.fetch.bind(window)
  const fetchMock = FetchMock.sandbox()
  globalThis.fetch = fetchMock as typeof globalThis.fetch

  fetchMock.config.overwriteRoutes = true
  fetchMock.mock(
    '*',
    (url, opts) =>
      new Promise((resolve: (result: Response) => void, reject) => {
        const request = new Request(url, opts)

        const matchedRule = matchNetworkRule(url, opts)
        if (!matchedRule) {
          realFetch(request).then((resp) => resolve(resp))
          return
        }

        const baseDetail = {
          requestType: 'fetch' as const,
          rule: matchedRule,
          request,
          resolve: withResolveScene(resolve),
          reject: (error: any = new TypeError('Failed to fetch')) =>
            reject(error),
        }
        const detail: NetworkBeforeEvent = {
          // id: new Date().getTime(),
          ...baseDetail,
          type: 'Run/network/before',
          pass: (intercept = false) => {
            passRequest(
              {
                ...baseDetail,
                type: 'Run/network/after',
              },
              {
                resolveResponse: resolve,
                intercept,
              }
            )
          },
        }
        onRun(detail)
      })
  )
  return () => resetFetch()
}

export const resetFetch = () => {
  if (!originalFetch) {
    console.warn('please setupFetch before resetFetch.')
    return
  }
  globalThis.fetch = originalFetch
  originalFetch = null
}
