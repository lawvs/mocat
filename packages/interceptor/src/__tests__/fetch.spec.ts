// polyfill jest fetch
// https://github.com/node-fetch/node-fetch
import * as nodeFetch from 'node-fetch'

import { setupFetch, resetFetch } from '../fetch'
import { eventEmitter, mockRoute } from '../eventEmitter'

beforeAll(() => {
  if (!globalThis.fetch) {
    ;(globalThis as any).fetch = nodeFetch
    ;(globalThis as any).Response = nodeFetch.Response
    ;(globalThis as any).Headers = nodeFetch.Headers
    ;(globalThis as any).Request = nodeFetch.Request
  }

  setupFetch()
})

afterAll(() => {
  resetFetch()
})

describe('fetch', () => {
  beforeAll(() => {
    mockRoute({ url: '/' })
  })

  test('should fetch works', async () => {
    const listener = jest.fn((payload) => payload.resolve({ response: 1 }))
    eventEmitter.once('Run/network/before', listener)
    const resp = await fetch('/')
    const data = await resp.text()

    expect(resp.status).toEqual(200)
    expect(data).toEqual('1')
    expect(listener).toBeCalledTimes(1)
  })

  test('should fetch works with status', async () => {
    const listener = jest.fn((payload) => payload.resolve({ status: 400 }))
    eventEmitter.once('Run/network/before', listener)
    const resp = await fetch('/')

    expect(resp.status).toEqual(400)
    expect(listener).toBeCalledTimes(1)
  })

  test('should fetch works with header', async () => {
    const listener = jest.fn((payload) =>
      payload.resolve({ headers: { status: 204 } })
    )
    eventEmitter.once('Run/network/before', listener)
    const resp = await fetch('/')

    expect(resp.headers.get('status')).toEqual('204')
    expect(listener).toBeCalledTimes(1)
  })

  test('should fetch works with empty response', async () => {
    const listener = jest.fn((payload) => payload.resolve({}))
    eventEmitter.once('Run/network/before', listener)
    const resp = await fetch('/')
    const data = await resp.text()

    expect(data).toEqual('')
    expect(listener).toBeCalledTimes(1)
  })

  test('should fetch works with json', async () => {
    const data = { code: 1, msg: 'success' }
    const listener = jest.fn((payload) => payload.resolve({ response: data }))

    eventEmitter.once('Run/network/before', listener)

    expect((await fetch('/')).json()).resolves.toEqual(data)
    expect(listener).toBeCalledTimes(1)
  })

  test('should fetch works when reject', async () => {
    const listener = jest.fn((payload) => payload.reject())
    eventEmitter.once('Run/network/before', listener)

    expect(fetch('/')).rejects.toThrowError('Failed to fetch')
    expect(listener).toBeCalledTimes(1)
  })
})

describe('snapshot test', () => {
  test('should register route match snapshot', () => {
    const listener = jest.fn((payload) => expect(payload).toMatchSnapshot())
    eventEmitter.on('Register/networkRoute', listener)
    mockRoute({
      name: 'login',
      desc: 'login api',
      url: '/api/login',
      scenes: [
        {
          name: 'login success',
          response: { code: 0, username: 'admin' },
        },
        {
          name: 'login failure',
          desc: 'password incorrect',
          status: 400,
          response: { code: 1, message: 'password incorrect' },
        },
      ],
    })
    expect(listener).toBeCalledTimes(1)
  })

  test('should network before event match snapshot', async () => {
    mockRoute({ url: '/' })
    const listener = jest.fn((payload) => expect(payload).toMatchSnapshot())
    eventEmitter.on('Run/network/before', listener)

    fetch('/')
    expect(listener).toBeCalledTimes(1)
  })
})
