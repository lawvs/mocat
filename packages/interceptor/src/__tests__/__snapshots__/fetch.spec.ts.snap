// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot test should network before event match snapshot 1`] = `
Object {
  "method": "*",
  "type": "Register/networkRoute",
  "url": "/",
}
`;

exports[`snapshot test should network before event match snapshot 2`] = `
Object {
  "pass": [Function],
  "reject": [Function],
  "request": Request {
    "agent": undefined,
    "compress": true,
    "counter": 0,
    "follow": 20,
    "size": 0,
    "timeout": 0,
    Symbol(Body internals): Object {
      "body": null,
      "disturbed": false,
      "error": null,
    },
    Symbol(Request internals): Object {
      "headers": Headers {
        Symbol(map): Object {},
      },
      "method": "GET",
      "parsedURL": Url {
        "auth": null,
        "hash": null,
        "host": null,
        "hostname": null,
        "href": "/",
        "path": "/",
        "pathname": "/",
        "port": null,
        "protocol": null,
        "query": null,
        "search": null,
        "slashes": null,
      },
      "redirect": "follow",
      "signal": null,
    },
  },
  "requestType": "fetch",
  "resolve": [Function],
  "rule": Object {
    "method": "*",
    "type": "Register/networkRoute",
    "url": "/",
  },
  "type": "Run/network/before",
}
`;

exports[`snapshot test should register route match snapshot 1`] = `
Object {
  "desc": "login api",
  "method": "*",
  "name": "login",
  "scenes": Array [
    Object {
      "name": "login success",
      "response": Object {
        "code": 0,
        "username": "admin",
      },
    },
    Object {
      "desc": "password incorrect",
      "name": "login failure",
      "response": Object {
        "code": 1,
        "message": "password incorrect",
      },
      "status": 400,
    },
  ],
  "type": "Register/networkRoute",
  "url": "/api/login",
}
`;
